scrapy/utils/datatypes.py

```diff
@@ -63,6 +79,46 @@ def pop(self, key, *args):
         return dict.pop(self, self.normkey(key), *args)
 
 
+class CaseInsensitiveDict(collections.UserDict):
+    """A dict-like structure that accepts strings or bytes as keys and allows case-insensitive lookups.
+
+    It also allows overriding key and value normalization by defining custom `normkey` and `normvalue` methods.
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        self._keys: dict = {}
+        super().__init__(*args, **kwargs)
+
+    def __getitem__(self, key: AnyStr) -> Any:
+        normalized_key = self.normkey(key)
+        return super().__getitem__(self._keys[normalized_key.lower()])
```

Comment:
Shouldn’t `.lower()` remain in `self.normkey()`? It feels like we are hardcoding key normalization here.

On the one hand, it suits the class name. Maybe what we should have is a parent class with most of the functionality and this class simply setting `normkey`? Or simply have the docstrings clarify that the caseinsensitive part is not customizable in subclasses.

On the other hand, since the argument not to worry too much about backward incompatibilities with the previous class was that it was not documented, maybe we should make this class private, and if so maybe we do not need to offer facilities for subclassing.

Reply:
The intention is to hardcode the normalization of the access, while delegating the normalization of the keys/values themselves to `normkkey/normvalue`. Consider the following example:

```python
In [1]: from scrapy.utils.datatypes import CaseInsensitiveDict

In [2]: CaseInsensitiveDict({"FOO": "bar"})  # key remains uppercase
Out[2]: <CaseInsensitiveDict: {'FOO': 'bar'}>

In [3]: CaseInsensitiveDict({"FOO": "bar"})["foo"]  # access via lowercase key
Out[3]: 'bar'

In [4]: class TitleCaseCaseInsensitiveDict(CaseInsensitiveDict):
   ...:    def normkey(self, key):
   ...:        return key.title()
   ...: 

In [5]: TitleCaseCaseInsensitiveDict({"FOO": "bar"})  # key is converted to title case
Out[5]: <TitleCaseCaseInsensitiveDict: {'Foo': 'bar'}>

In [6]: TitleCaseCaseInsensitiveDict({"FOO": "bar"})["foo"]  # access via lowercase key
Out[6]: 'bar'
```

My plan was to override this in the new `Headers` implementation, to go from `header-name` to `Header-Name`, but that probably only needs to be an implementation detail and not part of the public interface (in fact, we might not event need or want to do that, if Twisted capitalizes headers on its own). For now I'll rename to `_normkey/_normvalue`, we might even remove them altogether later.